monkey/monkey_island/cc/ui/src/components/ui-components/AdvancedMultiSelect.js

```diff
@@ -1,121 +1,204 @@
-            <Form.Group
-              key={i}
-              onClick={() => setPaneInfo(schema.items.$ref, registry, value, setInfoPaneParams)}>
-              <Button value={value} variant={'link'} disabled={disabled}
-                      onClick={() => onChange(getSelectValuesAfterClick(selectValue, value))}>
-                <FontAwesomeIcon icon={selectValue.includes(value) ? faCheckSquare : faSquare}/>
-              </Button>
-              <span className={'option-text'}>
-                {label}
-              </span>
-            </Form.Group>
-          );
-        })}
-      </Form.Group>
-      <InfoPane title={infoPaneParams.title} body={infoPaneParams.content} link={infoPaneParams.link}/>
-    </div>
-  );
+  onMasterCheckboxClick = () => {
+    if (this.state.masterCheckboxState === MasterCheckboxState.ALL) {
+      var newValues = [];
+    } else {
+      newValues = this.enumOptions.map(({value}) => value);
+    }
+
+    this.props.onChange(newValues);
+    this.setMasterCheckboxState(newValues);
+    this.setHideResetState(newValues);
+    this.setPaneInfoToDefault(this.unsafeOptionsSelected(newValues));
+  }
+
+  onChildCheckboxClick = (value) => {
+    let selectValues = this.getSelectValuesAfterClick(value);
+    this.props.onChange(selectValues);
+
+    this.setMasterCheckboxState(selectValues);
+    this.setHideResetState(selectValues);
+  }
+
+  getSelectValuesAfterClick(clickedValue) {
+    const valueArray = cloneDeep(this.props.value);
+
+    if (valueArray.includes(clickedValue)) {
+      return valueArray.filter(e => e !== clickedValue);
+    } else {
+      valueArray.push(clickedValue);
+      return valueArray;
+    }
+  }
+
+  setMasterCheckboxState(selectValues) {
+    this.setState(() => ({
+      masterCheckboxState: this.getMasterCheckboxState(selectValues)
+    }));
+  }
+
+  getMasterCheckboxState(selectValues) {
+    if (selectValues.length === 0) {
+      return MasterCheckboxState.NONE;
+    }
+
+    if (selectValues.length != this.enumOptions.length) {
+      return MasterCheckboxState.MIXED;
+    }
+
+    return MasterCheckboxState.ALL;
+  }
+
+  onResetClick = () => {
+    this.props.onChange(this.defaultValues);
+    this.setHideResetState(this.defaultValues);
+    this.setMasterCheckboxState(this.defaultValues);
+    this.setPaneInfoToDefault(this.unsafeOptionsSelected(this.defaultValues));
+  }
+
+  setHideResetState(selectValues) {
+    this.setState(() => ({
+      hideReset: this.getHideResetState(selectValues)
+    }));
+  }
+
+  getHideResetState(selectValues) {
+    return !(this.unsafeOptionsSelected(selectValues))
+  }
+
+  unsafeOptionsSelected(selectValues) {
+    return !(selectValues.every((value) => this.isSafe(value)));
+  }
+
+  isSafe = (itemKey) => {
+    return getFullDefinitionByKey(this.infoPaneRefString, this.registry, itemKey).safe;
+  }
+
+  setPaneInfo = (itemKey) =>  {
+    let definitionObj = getFullDefinitionByKey(this.infoPaneRefString, this.registry, itemKey);
```

Comment:
Maybe it would make sense to pull full objects in the constructor and save them in the state? Then you wouldn't even need `isSafe` method, you could just check `if (plugin.safe)`. And `getFullDefinitionByKey` would only be called once for each plugin.

Reply:
I could, but since the whole component uses the `value` string to identify checkboxes, I need to set up a map in the constructor that maps value strings to objects.

Then it would look like `if (this.options[itemKey].safe)`. For readability, I could change the `isSafe()` implementation to use that map. Additionally, the `ChildCheckboxContainer` makes calls to `isSafe()`.

We can marginally improve the performance, but I think `isSafe()` should stay.